src/ring_theory/graded_algebra/homogeneous_ideal.lean

```diff
@@ -0,0 +1,335 @@
+⟨⟨⊥, begin
+    intros i r hr, simp only [ideal.mem_bot] at hr,
+    rw [hr, alg_equiv.map_zero, zero_apply],
+    apply ideal.zero_mem
+  end⟩⟩
+
+instance homogeneous_ideal.has_top :
+  has_top (homogeneous_ideal A) :=
+⟨⟨⊤, λ _ _ _, by simp only [submodule.mem_top]⟩⟩
+
+@[simp] lemma homogeneous_ideal.eq_top_iff
+  (I : homogeneous_ideal A) : I = ⊤ ↔ I.1 = ⊤ :=
+⟨ λ h, by { rw h, refl },
+  λ h, begin
+    have h' : I.val = (⊤ : homogeneous_ideal A).val,
+    rw h, refl,
+    apply subtype.val_injective h',
+  end ⟩
+
+instance homogeneous_ideal.has_le : has_le (homogeneous_ideal A) :=
+{ le := λ I J, I.1 ≤ J.1 }
+
+instance homogeneous_ideal.order : preorder (homogeneous_ideal A) :=
+{ le_refl := λ I, by {unfold has_le.le, exact le_refl I.1, },
+  le_trans := λ I J K HIJ HJK,
+    by {unfold has_le.le at HIJ HJK ⊢, exact @le_trans _ _ I.1 J.1 K.1 HIJ HJK, },
+  ..(homogeneous_ideal.has_le A)}
+
+instance homogeneous_ideal.has_mem : has_mem R (homogeneous_ideal A) :=
+{ mem := λ r I, r ∈ I.1 }
+
+
+instance homogeneous_ideal.has_inf : has_inf (homogeneous_ideal A) :=
+{ inf := λ I J, begin
+  refine ⟨I.1 ⊓ J.1, _⟩,
+  intros i r hr, split, apply I.2 hr.1, apply J.2 hr.2,
+end }
+
+instance homogeneous_ideal.has_Inf  :
+  has_Inf (homogeneous_ideal A) :=
+{ Inf := λ ℐ, ⟨Inf (set.image (λ x : homogeneous_ideal A, x.val) ℐ), begin
+    intros i x Hx, simp only [submodule.mem_Inf] at Hx ⊢,
+    intros J HJ, simp only [set.mem_image, subtype.val_eq_coe] at HJ,
+    obtain ⟨K, HK₁, HK₂⟩ := HJ, rw ←HK₂,
+    have HK₃ := K.2,
+    apply HK₃, apply Hx, simp only [set.mem_image, subtype.val_eq_coe], use K, exact ⟨HK₁, rfl⟩,
+end⟩ }
+
+
+variable [Π (i : ι) (x : A i), decidable (x ≠ 0)]
+
+instance homogeneous_ideal.has_mul :
+  has_mul (homogeneous_ideal A) :=
+{ mul := λ I J, ⟨I.1 * J.1, begin
+    have HI := I.2, have HJ := J.2,
+    rw is_homogeneous_ideal_iff_exists at HI HJ,
+    choose s₁ hI using HI,
+    choose s₂ hJ using HJ,
+    rw is_homogeneous_ideal_iff_exists,
+    rw [hI, hJ, ideal.span_mul_span'],
+    refine ⟨s₁ * s₂, _⟩,
+    apply congr_arg,
+    ext, split; intro hx,
+    { rw set.mem_mul at hx,
+      obtain ⟨y1, y2, h1, h2, h3⟩ := hx,
+      rw set.mem_image at h1, obtain ⟨z1, h1⟩ := h1,
+      have hy1 : y1 ∈ set_like.homogeneous_submonoid A,
+      rw ←h1.2, exact z1.2,
+      rw set.mem_image at h2, obtain ⟨z2, h2⟩ := h2,
+      have hy2 : y2 ∈ set_like.homogeneous_submonoid A,
+      rw ←h2.2, exact z2.2,
+
+      use y1 * y2, apply submonoid.mul_mem,
+      exact hy1, exact hy2,
+      refine ⟨_, h3⟩, rw set.mem_mul, use y1, assumption,
+      use y2, assumption, tidy, },
+    { rw set.mem_image at hx,
+      obtain ⟨y, hy1, hy⟩ := hx,
+      rw set.mem_mul at hy1 ⊢,
+      obtain ⟨z1, z2, hz1, hz2, hz3⟩ := hy1,
+      use z1, use z2, split, rw set.mem_image, use z1, refine ⟨hz1, rfl⟩,
+      split, rw set.mem_image, use z2, refine ⟨hz2, rfl⟩, tidy, }
+  end⟩ }
```

Comment:
Given you have `is_homogenous_ideal` as a definition already, I think you should extract this proof to a lemma `is_homogenous_ideal.mul`. Same for all of these instances.